### ISSUE TITLE
Rounding issues - when will we finally be precise?

### ISSUE THREAD

**Issue by krychla1** [ccxt/ccxt#16167]
Dear fellows,
I am more and more desperate with rounding issues in ccxt.
I even went the hard way of researching **exchange.number: String** switch because of it, which has been suggested by @frosty00, which caused me even more pain and lost funds.  What should I do? Again and again I face rounding errors which can't be fixed in the current state of your framework. 

**Shouldn't **PRECISE MATH** be priority n1 when dealing with money??**

- OS: linux
- Programming Language version: nodejs
- CCXT version: 2.4.47

```JavaScript
//okex
let balance = exchange.fetchBalance({type: 'funding'});
//{"availBal":"828810596.00089356","bal":"828810596.00089356","ccy":"SOS","frozenBal":"0"}
console.log(balance['SOS'])
//{ free: 828810596.0008936, used: 0, total: 828810596.0008936 }
```

Don't tell me to use 'number': String switch please, unless you are ready to assist with problems using it (not just laugh in the eyes by closing it like #15238)

Thanks!

**Comment by samgermain** [comment]
This is something that you and I both want. Thank you for you patience with it. I have a project with all the string math PRs that I have opened https://github.com/orgs/ccxt/projects/52/views/1

Those PRs use string math for all the calculations involving money, but recently I've found out that there are some bugs related to math involving any value returned from the exchange if `'number': String` is used, so once all those PRs are merged, I'll need to do a final scan through the library for any arithmetic and take care of those bugs.

The updates are already written, but they need to be reviewed and merged, but there's a lot of PRs for this library, and only so many people that can review and merge them, so resources are limited. I'm pretty sure that one of the lead developers plans to go through all of them soon though. If you would like, you can go to the [ccxt discord server](https://discord.gg/ccxt), and try to talk to one of the members with the role **lead**. Those members are the ones who are able to review and merge PRs.

Sorry for the wait, and once again thank you for your patience.

**Comment by ttodua** [comment]
in addition to string math, there is also another path of issue, which are precisions per currency, which needs fetchCurrencies method to have full data available from exchange, to know the precisions for currencies (i am involved in that project, to work precisions/currencies/networks) and I have to admit, it will take a while until we cover that issue fully (I have started from most important exchanges and only after then considering lower end exchanges).

**Comment by krychla1** [comment]
@samgermain thank you very much for all the info. I came across the PRs someday, however I am not familiar how that works. E.g. for okx there is only okx.createOrder market buy string math #14966, however I cannot see any fix for fetchCurrencies (#15238), which I understand to be a blocker for strings used in okx exchange.

@ttodua that sounds promising, however users might do that on their own based on info from exchanges. Don't get me wrong, I appreciate every nice to have feature, but at current point we struggle to get correct essential info from exchanges (which is also prerequisite for the currency precision I suppose).

**Comment by samgermain** [comment]
> @samgermain thank you very much for all the info. I came across the PRs someday, however I am not familiar how that works. E.g. for okx there is only okx.createOrder market buy string math #14966, however I cannot see any fix for fetchCurrencies (#15238), which I understand to be a blocker for strings used in okx exchange.
> 
> @ttodua that sounds promising, however users might do that on their own based on info from exchanges. Don't get me wrong, I appreciate every nice to have feature, but at current point we struggle to get correct essential info from exchanges (which is also prerequisite for the currency precision I suppose).

I'll need to fix that too then. Once all the PRs in that project are merged, I'll do another scan of the library to find any arithmetic performed on numbers from exchange responses (which I'm guessing there won't be a lot, because I already went through and replaced all the money based arithemetic), and then we'll be ready to start using Precise and fixing individual bugs as they show up

If you notice a change you can tag me in an issue, or post a PR

- numbers need to be replaced with the string version of numbers
- `+`, `-`, `/`, `*`, `=`, `>`, `>=`, `<`, `<=` need to be replaced with `Precise.stringAdd`, `Precise.stringSub`, `Precise.stringDiv`, `Precise.stringMul`, `Precise.stringEq`, `Precise.stringGt`, `Precise.stringGE`, `Precise.stringLt`, `Precise.stringLE`
- `Math.max` and other `Math` need to be replaced with `Precise.stringMax` ...
- there are some exceptions to these rules, like math performed on the `since` argument, timestamps, simple counters ...
- string math must be used on all money calculations and numbers obtained from exchange responses

**Comment by ttodua** [comment]
@krychla1 please copy your message from https://github.com/ccxt/ccxt/issues/15238#issue-1402396605 and integrate into the top-post of this issue, so we can close that as duplicate, and leave this topic open.

**Comment by samgermain** [comment]
> @samgermain thank you very much for all the info. I came across the PRs someday, however I am not familiar how that works. E.g. for okx there is only okx.createOrder market buy string math #14966, however I cannot see any fix for fetchCurrencies (#15238), which I understand to be a blocker for strings used in okx exchange.
> 
> @ttodua that sounds promising, however users might do that on their own based on info from exchanges. Don't get me wrong, I appreciate every nice to have feature, but at current point we struggle to get correct essential info from exchanges (which is also prerequisite for the currency precision I suppose).

Can you write an [MRE](https://stackoverflow.com/help/minimal-reproducible-example) that I could use to reproduce the problem in OKX?

**Comment by krychla1** [comment]
@samgermain I can't create simpler example than the first one I posted
```Javascript 
//okex
let balance = exchange.fetchBalance({type: 'funding'});
//{"availBal":"828810596.00089356","bal":"828810596.00089356","ccy":"SOS","frozenBal":"0"}
console.log(balance['SOS'])
//{ free: 828810596.0008936, used: 0, total: 828810596.0008936 }
```
Obviously it depends on the balances you fetch, but the problem is not on the fetch function, it's on the rounding operation performed after. You can't round balances up, it creates mess.

**Comment by samgermain** [comment]
@krychla1 this problem was fixed with okx fetchCurrencies, thanks for letting us know and please let us know if you find any more cases where precise math needs to be implemented
